docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md

```diff
@@ -0,0 +1,236 @@
+  * To be found in the answer section:
+```
+   good-cname-2         CNAME good-cname-2-target
+   good-cname-2-target  A     127.0.0.1
+   good-cname-2-target  A     127.0.0.2
+```
+
+### GOOD-CNAME-CHAIN
+The query name will resolve to two `A` record via three CNAME.
+
+* Query name: "good-cname-chain.cname.recursor.engine.xa"
+  * To be found in the answer section:
+```
+   good-cname-chain         CNAME good-cname-chain-two
+   good-cname-chain-two     CNAME good-cname-chain-three
+   good-cname-chain-three   CNAME good-cname-chain-target
+   good-cname-chain-target  A     127.0.0.1
+```
+
+### GOOD-CNAME-OUT-OF-ZONE
+The query name will resolve to an `A` record via a CNAME in the zone with a
+target that points at a subzone, and a delegation to that sub zone.
+
+* For all queries
+  * The two zones, `cname.recursor.engine.xa` and
+    `goodsub.cname.recursor.engine.xa`, are hosted on different IP addresses.
+  * "x", "y" and "z" in the IP addresses in the configuration below are to be
+    set in the configuration of the test zones.
+
+* Query name: "good-cname-out-of-zone.cname.recursor.engine.xa"
+  * Servers: NS of `cname.recursor.engine.xa`.
+  * Answer, authority and additional sections, respectively, to be found in the
+    response:
+```
+   ;; ANSWER SECTION:
+   good-cname-out-of-zone  CNAME target.goodsub
+
+   ;; AUTHORITY SECTION:
+   goodsub                 NS    ns1.goodsub
+
+   ;; ADDITIONAL SECTION:
+   ns1.goodsub             A     127.x.y.z
+   ns1.goodsub             AAAA  fda1:b2:c3::127:x:y:z
+```
+* Query name: "target.goodsub.cname.recursor.engine.xa"
+  * Servers: NS of `goodsub.cname.recursor.engine.xa`.
+  * To be found in the answer section:
+```
+   target.goodsub          A     127.0.0.1
+```
+
+### NXDOMAIN-VIA-CNAME
+The query name exists, but as CNAME record. The target name of CNAME does not
+exist.
+
+* Query name: "nxdomain-via-cname.cname.recursor.engine.xa"
+  * The [RCODE Name] in the response is NxDomain.
+  * The target of the CNAME, `nxdomain-via-cname-target` does not exist in the,
+    zone and is not delegated.
+  * To be found in the answer section:
+```
+   nxdomain-via-cname       CNAME nxdomain-via-cname-target
+```
+
+### NODATA-VIA-CNAME
+The query name exists, but as CNAME record. The target name of CNAME does not
+exist.
+
+* Query name: "nodata-via-cname.cname.recursor.engine.xa"
+  * The target of the CNAME, `nodata-via-cname-target`, exists in the zone but
+    has neither `A` or `CNAME` record, and is not delegated.
+  * To be found in the answer section:
+```
+   nodata-via-cname       CNAME nodata-via-cname-target
+```
+
+### MULT-CNAME
+The query name exists, but as CNAME, as two CNAME records.
+
+* Query name: "mult-cname.cname.recursor.engine.xa"
+  * To be found in the answer section:
+```
+   mult-cname            CNAME mult-cname-target-1
+   mult-cname            CNAME mult-cname-target-2
+   mult-cname-target-1   A     127.0.0.1
+   mult-cname-target-2   A     127.0.0.2
```

Comment:
I think the response may or may not contain the A records.

After experimenting with cases of illegal uses of CNAME like this one in IBDNS, I came to the conclusion that I can't have IBDNS follow aliases in those cases. In fact, you’d hit problem cases that are similar to the reason why DNS query packets never contain more than one question: what if `mult-cname-target-1` exists but `mult-cname-target-2` doesn’t? What RCODE do you set in the response header?

So I suggest that the answer section should contain the two CNAME resource records, but without expecting the A records. Either way, it’ll be an illegal answer.

Reply:
After sleeping over it, I think you’re right and there is no need to change anything; please disregard my comment.

I was just wondering how an intentionally broken DNS server could systematically respond to zone files containing such illegal aliases when there are multiple CNAMEs for the same owner name. But this is not the right place to discuss this sort of thing.